tokens.get() incorrectly returns null in some cases

johncrim opened the pull request:
![](https://source.unsplash.com/featured/?cute,animal&CHANGE_ME)

## Description

fix(null for existing token): Handle color references in token values

Previously, `tokens.get()` incorrectly returned null depending on the value of the token. This occurred when the token value contained a hash (#) followed by text. Now, the hash and text remain in the token value if there is no token matching an #alias reference. This handles cases like `#rrggbb` within an unquoted string.

The previous behavior was a regression from accoutrement 3.x, so I don't think the docs or changelog need updating for this fix.

fixes: #99


## Steps to test/reproduce

New tests included in the PR to repro the issue. If you revert the changes, the new tests fail.

mirisuzanne commented on sass/tokens/_parse.scss:
```diff
@@ -232,28 +251,43 @@ $_stack: ();
+      // full alias (eg value is '#foo->not-here').
       @if (meta.type-of($lookup) != 'map') {
-        @return $lookup;
+        @if (1 == map.get($alias, 'start')) and
+          (string.length($alias-key) + 2 == map.get($alias, 'end'))
+        {
```
Could this be a comparison of `$value == '##{$alias-key}'`? That seems a bit more clear and to the point?

johncrim replied:
Yes, I agree. Will update. Thanks!

mirisuzanne commented on sass/tokens/_parse.scss:
```diff
@@ -232,28 +251,43 @@ $_stack: ();
-      // this is a recursive dependency
-      $new: compile-token($map, map.get($lookup, 'token'), $alias);
+      // Replace #alias when lookup succeeds. Otherwise leave it in the string (eg #aaa)
+      @if (meta.type-of($lookup) == 'map') {
```
One of these could be an `@else`? If we reverse the order, we could handle the lookup failure with `@else if`?

johncrim replied:
Done

mirisuzanne commented on test/tokens/_api.scss:
```diff
@@ -131,8 +131,27 @@
+    }
+
+    @include it('Handles CSS values in maps') {
+      $gradients: (
+        'accent':
+          linear-gradient(hsl(31deg 94% 52%) 10%, hsl(274deg 95% 57%) 80%),
+      );
```
lets clarify this by using hex notation for the colors, and updating the test description so it explains the issue a bit more?

johncrim replied:
Sounds good - will do.

mirisuzanne reviewed (commented):
Yeah, I think your solution here makes a lot of sense. Thanks! I left a  couple questions, if you have time to address those.

johncrim commented:
Hi @mirisuzanne  - please take another look when you have time. 

After addressing your feedback, I realized I could make another simplification - `next-alias()` doesn't need to include the `start` key, since it is no longer used. We could put it back in if you think there's any other utility for having the start index.

mirisuzanne commented on test/tokens/_api.scss:
```diff
@@ -131,8 +131,23 @@
+
+    @include it('Allows CSS values containing hashes') {
+      $gradients: (
+        'accent': linear-gradient(#f88812 10%, #9f29fa 80%),
```
Interesting side-note: Sass currently treats unknown/css functions as strings, which is part of the issue in this specific case. We're talking about what it might mean to add a CSS function data type…

johncrim replied:
Very interesting indeed - thanks for the insight.

Perhaps related: since unquoted strings have to be valid CSS they behave quite differently from quoted strings. I came up against the fact that there is no `string.is-quoted()` fn in sass:string, which seems like an oversight; but then I decided I didn't need it.

mirisuzanne replied:
@johncrim Can you file an issue on the Sass repo? I think it would be good to document some of the use-cases for it. Now that we have modules, we're looking into what functions are missing from them.

mirisuzanne replied:
oh, I guess if you didn't need it, maybe there's nothing there?

johncrim replied:
Hi @mirisuzanne  - there's a [long-standing issue asking for `string.is-quoted()`](https://github.com/sass/sass/issues/1291).

I'll add a comment to that issue.

mirisuzanne reviewed (approved):
looks great! I agree the start value isn't necessary here. These functions are private, so if we don't need it internally, there's not much reason to add it.
